### PR TITLE
Add metric cache and migrate existing metrics to cache implementation

### LINKED ISSUE #2140: Metrics Caching
### 🚀 The feature

Build two hashmap to cache the metrics defined in yaml config file. One is used to cache frontend metrics; the other one is used to cache backend metrics.

### Motivation, pitch

See details in #1492

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lxning**
## Description
Implementation of Metric Cache and migration of existing metrics to cache implementation
Design: #1492 

Fixes #2140 #2141 

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Feature/Issue validation/testing
- [ ] Unit tests
Included in PR and run as part of sanity test in CI

- [ ] Manual end to end tests to verify metrics frontend implementation
Verified both log and prometheus modes

- [ ] Regression tests
    - CPU: [cpu_regression_log.txt](https://github.com/pytorch/serve/files/11314991/cpu_regression_log.txt)
        - Relevant passing tests:
            - test_metrics.py::test_metrics_log_mode
            - test_metrics.py::test_metrics_prometheus_mode
        - Unrelated failing tests:
            - test_handler.py::test_echo_stream_inference
    - GPU: [gpu_regression_log.txt](https://github.com/pytorch/serve/files/11314997/gpu_regression_log.txt)
        - Relevant passing tests:
            - test_metrics.py::test_metrics_log_mode
            - test_metrics.py::test_metrics_prometheus_mode
        - Unrelated failing tests:
            - test_distributed_inference_handler.py::test_large_model_inference
            - test_handler.py::test_echo_stream_inference
            - test_sm_mme_requirements.py::test_oom_on_model_load
            - test_sm_mme_requirements.py::test_oom_on_invoke

- [ ] Benchmark
    - CPU: [cpu_ab_report.csv](https://github.com/pytorch/serve/files/11315047/cpu_ab_report.csv)
    - GPU: [gpu_ab_report.csv](https://github.com/pytorch/serve/files/11315048/gpu_ab_report.csv)

## Checklist:

- [x] Did you have fun?
- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerLifeCycle.java [hunk 1]
```diff
@@ -296,7 +297,33 @@ public void run() {
...
                         Metric parsedMetric = Metric.parse(matcher.group(3));
                         if (parsedMetric != null) {
-                            loggerModelMetrics.info(parsedMetric.toString());
+                            if (METRIC_CACHE.getMetricBackend(parsedMetric.getMetricName())
+                                    != null) {
+                                try {
+                                    List<String> dimensionValues = new ArrayList<String>();
+                                    for (Dimension dimension : parsedMetric.getDimensions()) {
+                                        dimensionValues.add(dimension.getValue());
+                                    }
```
Comment (lxning) [inline frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerLifeCycle.java:306]: why not directly parse the backend log to generate IMetric?

File: frontend/server/src/main/java/org/pytorch/serve/metrics/MetricCollector.java [hunk 2]
```diff
@@ -79,7 +79,32 @@ public void run() {
                     if (metric == null) {
                         logger.warn("Parse metrics failed: " + line);
                     } else {
-                        loggerMetrics.info("{}", metric);
+                        if (METRIC_CACHE.getMetricFrontend(metric.getMetricName()) != null) {
```
[the remark this answers is not in the document]
Comment (namannandan) [reply frontend/server/src/main/java/org/pytorch/serve/metrics/MetricCollector.java]: Yes, this is scheduled and run by the MetricManager: https://github.com/pytorch/serve/blob/master/frontend/server/src/main/java/org/pytorch/serve/metrics/MetricManager.java#L23-L27

**Review by lxning — COMMENTED** [review]
testModelMetrics should be updated and reflected in integration test ModelServerTest.java

**Review by msaroufim — APPROVED** [review]
Approving after chatting about this PR offline, I would still like to see some more docs around how a user goes about creating a custom python metric and how everything pipes back into a prometheus dashboard, critical to mention this in release notes as well and a tutorial as well although for the tutorial I'm fine if we do it in a followup PR

**Review by agunapal — APPROVED** [review]

**Pull request merged**